src/coreclr/jit/emit.cpp

```diff
@@ -4102,6 +4122,315 @@ void emitter::emitDispCommentForHandle(size_t handle, GenTreeFlags flag)
 #endif // DEBUG
 }
 
+/*****************************************************************************
+ *  Checks all jumps in the jump list to see if they are unconditional jumps
+ * generated at the end of BB_ALWAYS basic blocks and whether they are now a jump
+ * to the next instruction, this can happen when the following basic block emitted
+ * no instructions
+ *
+ * returns true if any jumps have been removed, otherwise returns false
+ */
+bool emitter::emitRemoveJumpToNextInst()
+{
+#ifdef TARGET_XARCH
+#ifdef DEBUG
+    if (emitComp->verbose)
+    {
+        printf("*************** In emitRemoveJumpToNextInst()\n");
+    }
+    if (EMIT_INSTLIST_VERBOSE)
+    {
+        printf("\nInstruction list before unconditional jump to next instruction removal:\n\n");
+        emitDispIGlist(true);
+    }
+    if (EMITVERBOSE)
+    {
+        emitDispJumpList();
+    }
+#endif
+
+    instrDescJmp*  jmp;
+    UNATIVE_OFFSET adjIG;
+    UNATIVE_OFFSET adjLJ;
+    insGroup*      lstIG;
+    int            removedCount   = 0;
+    UNATIVE_OFFSET removedSize    = 0;
+    int            iterationCount = 1;
+    bool           jmpRemoved     = false;
+
+/*****************************************************************************/
+/* If we iterate to look for more jumps to remove, we start again here.     */
+/*****************************************************************************/
+
+AGAIN:
+
+    lstIG                     = nullptr;
+    adjLJ                     = 0;
+    adjIG                     = 0;
+    jmp                       = emitJumpList;
+    instrDescJmp* previousJmp = nullptr;
+
+    while (jmp)
+    {
+        // if the jump is unconditional then it is a candidate
+        if (jmp->idInsFmt() == IF_LABEL && emitIsUncondJump(jmp) && !jmp->idjKeepLong && jmp->idGetJmpAlwaysFlag())
+        {
+            // unset the flag so that we know it has been processed already
+            jmp->idSetJmpAlwaysFlag(0);
+
+            // target group is not bound yet so use the cookie to fetch it
+            insGroup* targetGroup = (insGroup*)emitCodeGetCookie(jmp->idAddr()->iiaBBlabel);
+
+            if (targetGroup != nullptr && jmp->idjIG->igNext == targetGroup)
+            {
+                insGroup* group            = jmp->idjIG;
```

Comment:
You will have to skip the `group` if it has alignment instruction in the end. 

Just for your context, loop alignment happens in 3 phases:
1. Before codegen, we would identify the blocks that will contain `align` instruction in the end. This is in `placeLoopAlignInstructions()`. This explicitly identifies blocks that ends with `BBJ_ALWAYS`. 
2. During codegen, inside `genCodeForBBList()`, we will insert `align` instruction in IG whose corresponding BB ends with `BBF_HAS_ALIGN`.
3. After codegen, inside `emitLoopAlignAdjustments()`, we go through all the align instructions and adjust their size to the actual alignment we will be emitting.

`emitRemoveJumpToNextInst()` happens between #2 and #3,  you will have to make sure that you don't take off the jumps that were selected by alignment phase to hide `align` instructions.

Reply:
Excellent information. There's no way to get this information from the ig so i'll have to track back to the bb to check the flags won't I?
How does this work with `emitJumpDistBind()` which changes the length of the jump instructions? it commonly reduces their size by up to 3 bytes and the alignment mechanism must already be able to cope with this, presumably by producing a nop of the required size to fill the newly available space as well as any that was already needed for alignment. If that's the case won't the loss of an instruction already be compensatable? or is the maximum amount of padding allowed decided when the ig is created?

so if you've got something like:
```
mov rax, rcx
jmp LABEL_NEXT
align
```
is the maximum length of align decided once the ig is saved? If not then couldn't the aligner deal with it being transformed to:
```
mov rax, rcx
align
```
by simply extending the alignment? or can the removal of the instruction cause the need to re-evaluate whether padding is needed at all?

Reply:
> here's no way to get this information from the ig so i'll have to track back to the bb to check the flags won't I?

You can check `IGF_HAS_ALIGN` flag that we set inside `emitLoopAlign()` (which is called from `genCodeForBBList()`.

> emitJumpDistBind

That is the reason of 3rd step of adjustment which happens after `emitJumpDistBind()`. That phases exclusively check for such size reduction and accordingly adjust the `align` needed.

>  is the maximum amount of padding allowed decided when the ig is created?

Correct. In step 2, we always create alignment of 15 bytes and then shrink them in step 3.

> couldn't the aligner deal with it 

Our preference is to hide the `align` instruction behind jump so they are not fetched to begin with. So, although it has an ability to adjust the size, the reason I am suggesting to filter such blocks is to have `align` hidden behind the jump.

See the diagram in https://github.com/dotnet/runtime/pull/60787.

Reply:
I think it's worth leaving a comment in the code describing why we want to keep the jmp when alignment instructions are present. It has a material effect and that information should be in the code. I'll do that next push if I can. 
I have implemented the align check and skip group if it's set logic.